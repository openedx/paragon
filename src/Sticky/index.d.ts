import React from 'react';

export interface StickyProps {
  children: React.ReactNode;
  position?: 'top' | 'bottom';
  offset?: number | string;
  className?: string;
}

declare const Sticky: React.ForwardRefExoticComponent<StickyProps>;

export default Sticky;
