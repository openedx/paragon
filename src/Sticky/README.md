---
title: 'Sticky'
type: 'component'
components:
- Sticky
categories:
- Content
status: 'New'
designStatus: 'Done'
devStatus: 'In progress'
notes: |
---

Makes elements remain at the top or bottom of the viewport, like a sticky navbar.

## Basic Usage

### Top positioning

```jsx live
<div style={{ display: 'flex', width: '100%', height: '400px', border: '2px dashed grey' }}>
  <Sticky>
    <div style={{ background: '#0000aa', width: '100%', height: '150px', color: '#ffffff'}}>
      Sticky to the top
    </div>
  </Sticky>
</div>
```

### Bottom positioning

```jsx live
<div style={{ display: 'flex', width: '100%', height: '400px', border: '2px dashed grey' }}>
  <Sticky position="bottom">
    <div style={{ background: '#aa0000', width: '100%', height: '150px', color: '#ffffff' }}>
      Sticky to the bottom
    </div>
  </Sticky>
</div>
```

### With offset

Valid `offset` values are the same as for [the spacing classes](/foundations/spacing).

```jsx live
<div style={{ display: 'flex', width: '100%', height: '400px', border: '2px dashed grey' }}>
  <Sticky offset={3.5}>
    <div style={{ background: '#00aa00', width: '100%', height: '150px', color: '#ffffff' }}>
      Sticky with offset
    </div>
  </Sticky>
</div>
```