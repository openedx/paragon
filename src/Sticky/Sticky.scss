@import "variables";

.pgn__sticky {
  display: flex;
  width: 100%;
  position: sticky;

  &.pgn__sticky-top {
    align-self: self-start;
    top: $sticky-offset;

    @each $level, $offset in $spacers {
      &.pgn__sticky-offset--#{$level} {
        top: $offset;
      }
    }

    &.pgn__sticky-shadow {
      box-shadow: $sticky-shadow-bottom;
    }
  }

  &.pgn__sticky-bottom {
    align-self: self-end;
    bottom: $sticky-offset;

    @each $level, $offset in $spacers {
      &.pgn__sticky-offset--#{$level} {
        bottom: $offset;
      }
    }

    &.pgn__sticky-shadow {
      box-shadow: $sticky-shadow-top;
    }
  }
}
