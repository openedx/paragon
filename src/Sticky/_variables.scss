$sticky-offset:                       0 !default;
$sticky-shadow-top:                   0 -.5rem 1rem rgba(0, 0, 0, .15), 0 -.25rem .625rem rgba(0, 0, 0, .15) !default;
$sticky-shadow-bottom:                0 .5rem 1rem rgba(0, 0, 0, .15), 0 .25rem .625rem rgba(0, 0, 0, .15) !default;
