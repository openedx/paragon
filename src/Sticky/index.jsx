import React, { useLayoutEffect, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

const POSITION_VARIANTS = [
  'top',
  'bottom',
];

const Sticky = React.forwardRef(({
  position,
  children,
  offset,
  className,
  ...rest
}, ref) => {
  const [isSticky, setIsSticky] = useState(false);
  const defaultRef = React.useRef();
  const resolvedRef = ref || defaultRef;

  // eslint-disable-next-line consistent-return
  useLayoutEffect(() => {
    if (resolvedRef.current) {
      const stickyElement = resolvedRef.current;
      // getComputedStyle is used to get real top/bottom
      // values on the page for proper shadows display
      const elementStyles = window.getComputedStyle(stickyElement);
      const elementOffset = elementStyles[position || 'top'];
      // Margin calculations according to the offset.
      // 1 pixel above/bellow + offset pixels that determines
      // when callback function is called
      const elementWithOffset = 1 + (parseInt(elementOffset, 10) || 0);
      const observer = new IntersectionObserver(
        ([entry]) => setIsSticky(entry.intersectionRatio < 1),
        {
          threshold: [1],
          rootMargin: position === 'bottom'
            ? `0px 0px -${elementWithOffset}px 0px`
            : `-${elementWithOffset}px 0px 0px 0px`,
        },
      );
      observer.observe(stickyElement);

      return () => {
        observer.unobserve(stickyElement);
      };
    }
  }, [position, resolvedRef]);

  return (
    <div
      className={classNames(
        'pgn__sticky',
        `pgn__sticky-${position || 'top'}`,
        offset ? `pgn__sticky-offset--${offset}` : '',
        { 'pgn__sticky-shadow': isSticky },
        className,
      )}
      ref={resolvedRef}
      {...rest}
    >
      {children}
    </div>
  );
});

Sticky.propTypes = {
  /** Specifies content of the component. */
  children: PropTypes.node.isRequired,
  /** Specifies position of the element. */
  position: PropTypes.oneOf(POSITION_VARIANTS),
  /**
   * Specifies offset from top/bottom depending on the `position` property.
   *
   * Valid values are based on `the spacing classes`:
   * `0, 0.5, ... 6`.
   */
  offset: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Specifies an additional `className` to add to the base element. */
  className: PropTypes.string,
};

Sticky.defaultProps = {
  position: 'top',
  offset: undefined,
  className: undefined,
};

export default Sticky;
