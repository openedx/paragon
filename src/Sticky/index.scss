:root {
  --pgn-elevation-sticky-shadow-bottom:
    var(--pgn-elevation-sticky-shadow-bottom-1-offset-x)
    var(--pgn-elevation-sticky-shadow-bottom-1-offset-y)
    var(--pgn-elevation-sticky-shadow-bottom-1-blur)
    var(--pgn-elevation-sticky-shadow-bottom-1-color),
    var(--pgn-elevation-sticky-shadow-bottom-2-offset-x)
    var(--pgn-elevation-sticky-shadow-bottom-2-offset-y)
    var(--pgn-elevation-sticky-shadow-bottom-2-blur)
    var(--pgn-elevation-sticky-shadow-bottom-2-color);
}

.pgn__sticky {
  display: flex;
  width: 100%;
  position: sticky;

  &.pgn__sticky-top {
    align-self: self-start;
    top: var(--pgn-spacing-sticky-offset);

    @each $level, $offset in $spacers {
      &.pgn__sticky-offset--#{$level} {
        top: $offset;
      }
    }

    &.pgn__sticky-shadow {
      box-shadow: var(--pgn-elevation-sticky-shadow-bottom);
    }
  }

  &.pgn__sticky-bottom {
    align-self: self-end;
    bottom: var(--pgn-spacing-sticky-offset);

    @each $level, $offset in $spacers {
      &.pgn__sticky-offset--#{$level} {
        bottom: $offset;
      }
    }

    &.pgn__sticky-shadow {
      box-shadow: var(--pgn-elevation-sticky-shadow-top);
    }
  }
}
