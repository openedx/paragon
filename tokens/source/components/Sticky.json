{
  "sticky": {
    "offset": { "value": "0", "source": "$sticky-offset" },
    "shadow": {
      "top": {
        "value": "0 -.5rem 1rem rgba(0, 0, 0, .15), 0 -.25rem .625rem rgba(0, 0, 0, .15)",
        "source": "$sticky-shadow-top"
      },
      "bottom": {
        "value": "0 .5rem 1rem rgba(0, 0, 0, .15), 0 .25rem .625rem rgba(0, 0, 0, .15)",
        "source": "$sticky-shadow-bottom"
      }
    }
  }
}
