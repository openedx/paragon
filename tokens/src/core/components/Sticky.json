{
  "spacing": {
    "$type": "dimension",
    "sticky": {
      "offset": {
        "$value": "0",
        "source": "$sticky-offset"
      }
    }
  }
}
