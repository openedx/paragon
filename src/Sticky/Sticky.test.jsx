import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';

import Sticky from './index';

const observe = jest.fn();
const unobserve = jest.fn();

window.IntersectionObserver = jest.fn(() => ({
  observe,
  unobserve,
}));

describe('<Sticky />', () => {
  describe('correct rendering', () => {
    it('renders without props', () => {
      const tree = renderer.create((
        <Sticky>Sticky</Sticky>
      )).toJSON();
      expect(tree).toMatchSnapshot();
    });
    it('renders with top positioning', () => {
      const wrapper = mount(<Sticky>content</Sticky>);
      const sticky = wrapper.find('.pgn__sticky');
      expect(sticky.hasClass('pgn__sticky-top')).toEqual(true);
      wrapper.setProps({ position: 'top' });
      expect(sticky.hasClass('pgn__sticky-top')).toEqual(true);
    });
    it('renders with bottom positioning', () => {
      const wrapper = mount(<Sticky position="bottom">content</Sticky>);
      const sticky = wrapper.find('.pgn__sticky');
      expect(sticky.hasClass('pgn__sticky-bottom')).toEqual(true);
    });
    it('renders with offset', () => {
      const offset = 3;
      const wrapper = mount(<Sticky offset={offset}>content</Sticky>);
      const sticky = wrapper.find('.pgn__sticky');
      expect(sticky.hasClass(`pgn__sticky-offset--${offset}`)).toEqual(true);
    });
    it('observer is initialized during render and detached during unmount', () => {
      const wrapper = mount(<Sticky>content</Sticky>);
      expect(observe).toHaveBeenCalled();
      wrapper.unmount();
      expect(unobserve).toHaveBeenCalled();
    });
  });
});
