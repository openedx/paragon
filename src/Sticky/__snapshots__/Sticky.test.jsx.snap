// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sticky /> correct rendering renders without props 1`] = `
<div
  className="pgn__sticky pgn__sticky-top"
>
  Sticky
</div>
`;
